compiler/qsc_partial_eval/src/lib.rs

```diff
@@ -0,0 +1,710 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#[cfg(test)]
+mod tests;
+
+use qsc_data_structures::functors::FunctorApp;
+use qsc_eval::{
+    self, backend::Backend, exec_graph_section, output::GenericReceiver, val::Value, Env, State,
+    StepAction, StepResult,
+};
+use qsc_fir::{
+    fir::{
+        Block, BlockId, CallableDecl, CallableImpl, ExecGraphNode, Expr, ExprId, ExprKind, Global,
+        LocalItemId, PackageId, PackageStore, PackageStoreLookup, Pat, PatId, SpecDecl, SpecImpl,
+        Stmt, StmtId, StmtKind, StoreBlockId, StoreExprId, StoreItemId, StorePatId, StoreStmtId,
+    },
+    ty::{Prim, Ty},
+    visit::Visitor,
+};
+use qsc_rca::{ComputeKind, ComputePropertiesLookup, PackageStoreComputeProperties, ValueKind};
+use qsc_rir::rir::{self, Callable, CallableId, CallableType, Instruction, Literal, Program};
+use rustc_hash::FxHashMap;
+use std::{rc::Rc, result::Result};
+
+pub fn partially_evaluate(
+    package_id: PackageId,
+    package_store: &PackageStore,
+    compute_properties: &PackageStoreComputeProperties,
+) -> Result<Program, Error> {
+    let partial_evaluator = PartialEvaluator::new(package_id, package_store, compute_properties);
+    partial_evaluator.eval()
+}
+
+pub enum Error {
+    EvaluationFailed(qsc_eval::Error),
+}
+
+struct PartialEvaluator<'a> {
+    package_store: &'a PackageStore,
+    compute_properties: &'a PackageStoreComputeProperties,
+    assigner: Assigner,
+    backend: QubitsAndResultsAllocator,
+    callables_map: FxHashMap<String, CallableId>,
```

Comment:
I think if you use `Rc<str>` here instead of `String` you can void some memory allocations and just `.clone()` the value from the FIR, which is faster than using `.to_string()`:

```suggestion
    callables_map: FxHashMap<Rc<str>, CallableId>,
```

Reply:
Updated.